File: stocksense-backend/utils/symbol_resolver.py
# Maps common names to NSE symbols
COMPANY_MAP = {
    "ADANI ENT": "ADANIENT",
    "ADANI ENTERPRISES": "ADANIENT",
    "ADANI PORTS": "ADANIPORTS",
    "AIRTEL": "BHARTIARTL",
    "BHARTI AIRTEL": "BHARTIARTL",
    "BAJAJ FINANCE": "BAJFINANCE",
    "BAJAJ FINSERV": "BAJAJFINSV",
    "CG POWER": "CGPOWER",
    "GLAND PHARMA": "GLAND",
    "HDFC BANK": "HDFCBANK",
    "HDFC LIFE": "HDFCLIFE",
    "HERO MOTO": "HEROMOTOCO",
    "HINDUSTAN UNILEVER": "HINDUNILVR",
    "HUL": "HINDUNILVR",
    "ICICI BANK": "ICICIBANK",
    "INFOSYS": "INFY",
    "ITC LTD": "ITC",
    "KOTAK BANK": "KOTAKBANK",
    "KOTAK MAHINDRA": "KOTAKBANK",
    "LARSEN": "LT",
    "L&T": "LT",
    "MAHINDRA": "M&M",
    "MARUTI SUZUKI": "MARUTI",
    "NESTLE": "NESTLEIND",
    "NTPC LTD": "NTPC",
    "ONGC LTD": "ONGC",
    "RELIANCE IND": "RELIANCE",
    "RELIANCE INDUSTRIES": "RELIANCE",
    "SBI": "SBIN",
    "STATE BANK": "SBIN",
    "STALLION IND": "STALLION",
    "SUN PHARMA": "SUNPHARMA",
    "TATA CONSUMER": "TATACONSUM",
    "TATA MOTORS": "TATAMOTORS",
    "TATA STEEL": "TATASTEEL",
    "ULTRA CEMENT": "ULTRACEMCO",
    "VADILAL IND": "VADILALIND",
    "WIPRO LTD": "WIPRO",
}

def resolve_symbol(raw_input: str) -> str:
    clean = raw_input.strip().upper()
    
    # Check company name map
    if clean in COMPANY_MAP:
        return COMPANY_MAP[clean]
    
    # Remove common suffixes
    for suffix in [
        ' LIMITED', ' LTD', ' LTD.',
        ' INC', ' CORP', '.NS', 
        '.BO', ':NSE', ':BSE'
    ]:
        clean = clean.replace(suffix, '')
    
    return clean.strip()

def format_for_twelve_data(symbol: str) -> str:
    resolved = resolve_symbol(symbol)
    return f"{resolved}:NSE"

def format_for_stooq(symbol: str) -> str:
    resolved = resolve_symbol(symbol)
    return f"{resolved.lower()}.in"

File: src/utils/symbolResolver.js
/**
 * symbolResolver.js
 * Universal Symbol Resolver for Twelve Data API and Stooq.com
 */

const SYMBOL_CORRECTIONS = {
  'ADANIENT': 'ADANIENT',
  'ADANIPORTS': 'ADANIPORTS', 
  'BAJAJFINSV': 'BAJAJFINSV',
  'BHARTIARTL': 'BHARTIARTL',
  'BPCL': 'BPCL',
  'BRITANNIA': 'BRITANNIA',
  'CGPOWER': 'CGPOWER',
  'CIPLA': 'CIPLA',
  'COALINDIA': 'COALINDIA',
  'DRREDDY': 'DRREDDY',
  'EICHERMOT': 'EICHERMOT',
  'GLAND': 'GLAND',
  'GRASIM': 'GRASIM',
  'HAL': 'HAL',
  'HCLTECH': 'HCLTECH',
  'HDFC': 'HDFCBANK',
  'HDFCBANK': 'HDFCBANK',
  'HDFCLIFE': 'HDFCLIFE',
  'HEROMOTOCO': 'HEROMOTOCO',
  'HINDALCO': 'HINDALCO',
  'HINDUNILVR': 'HINDUNILVR',
  'ICICIBANK': 'ICICIBANK',
  'INDUSINDBK': 'INDUSINDBK',
  'INFY': 'INFY',
  'ITC': 'ITC',
  'JIOFIN': 'JIOFIN',
  'JSWSTEEL': 'JSWSTEEL',
  'KOTAKBANK': 'KOTAKBANK',
  'LT': 'LT',
  'LTIM': 'LTIM',
  'MARUTI': 'MARUTI',
  'MM': 'M&M',
  'NESTLEIND': 'NESTLEIND',
  'NTPC': 'NTPC',
  'ONGC': 'ONGC',
  'POWERGRID': 'POWERGRID',
  'RELIANCE': 'RELIANCE',
  'SBIN': 'SBIN',
  'SHRIRAMFIN': 'SHRIRAMFIN',
  'STALLION': 'STALLION',
  'SUNPHARMA': 'SUNPHARMA',
  'SUZLON': 'SUZLON',
  'TATACONSUM': 'TATACONSUM',
  'TATAMOTORS': 'TATAMOTORS',
  'TATASTEEL': 'TATASTEEL',
  'TCS': 'TCS',
  'TECHM': 'TECHM',
  'TITAN': 'TITAN',
  'TRENT': 'TRENT',
  'ULTRACEMCO': 'ULTRACEMCO',
  'UPL': 'UPL',
  'VADILALIND': 'VADILALIND',
  'WIPRO': 'WIPRO',
  'ZOMATO': 'ZOMATO'
};

export const resolveSymbolForTwelveData = (input) => {
  if (!input) return 'RELIANCE:NSE';

  // Clean input
  const clean = input
    .trim()
    .toUpperCase()
    .replace(/\.NS$/i, '')
    .replace(/\.BO$/i, '')
    .replace(/\.NSE$/i, '')
    .replace(/\.BSE$/i, '')
    .replace(/:NSE$/i, '')
    .replace(/:BSE$/i, '')
    .replace(/\.IN$/i, '');

  // Check corrections map
  const corrected = SYMBOL_CORRECTIONS[clean] || clean;

  // Return Twelve Data format: SYMBOL:NSE
  return `${corrected}:NSE`;
};

export const resolveSymbolForStooq = (input) => {
  if (!input) return 'reliance.in';

  const clean = input
    .trim()
    .toUpperCase()
    .replace(/\.NS$/i, '')
    .replace(/\.BO$/i, '')
    .replace(/\.NSE$/i, '')
    .replace(/:NSE$/i, '')
    .replace(/\.IN$/i, '');

  const corrected = SYMBOL_CORRECTIONS[clean] || clean;

  // Stooq format: reliance.in (lowercase)
  return `${corrected.toLowerCase()}.in`;
};
